File: src/epics_robot/src/add_stage_to_scene.cpp
#include <rclcpp/rclcpp.hpp>

#include <moveit/planning_scene_interface/planning_scene_interface.h>

#include <moveit_msgs/msg/collision_object.hpp>
#include <moveit_msgs/msg/planning_scene.hpp>
#include <moveit_msgs/msg/allowed_collision_matrix.hpp>
#include <moveit_msgs/msg/allowed_collision_entry.hpp>
#include <moveit_msgs/srv/get_planning_scene.hpp>
#include <moveit_msgs/srv/apply_planning_scene.hpp>
#include <shape_msgs/msg/mesh.hpp>
#include <geometry_msgs/msg/pose.hpp>

#include <geometric_shapes/shape_operations.h>   // shapes::createMeshFromResource, constructMsgFromShape
#include <geometric_shapes/mesh_operations.h>

#include <chrono>
#include <thread>
#include <string>
#include <vector>

using namespace std::chrono_literals;

class StageSceneSetup : public rclcpp::Node
{
public:
  StageSceneSetup()
  : Node("stage_scene_setup")
  {
    RCLCPP_INFO(get_logger(), "Stage Scene Setup node initialized");
  }

  bool add_stage_mesh()
  {
    // 1) Remove previous objects (optional)
    RCLCPP_INFO(get_logger(), "Clearing previous scene objects...");
    auto names = planning_scene_interface_.getKnownObjectNames();
    if (!names.empty())
      planning_scene_interface_.removeCollisionObjects(names);

    std::this_thread::sleep_for(500ms);

    // 2) Build CollisionObject with Mesh
    const std::string object_id = "stage";
    const std::string frame_id  = "base_link";

    // IMPORTANT:
    // createMeshFromResource expects a resource URI.
    // For a plain absolute path, use "file://".
    const std::string stl_resource = "file:///home/stevek/ws/stage.stl";

    // Scale (convert cm to meter: 0.01)
    const Eigen::Vector3d scale(0.01, 0.01, 0.01);

    shapes::Mesh* mesh = shapes::createMeshFromResource(stl_resource, scale);
    if (!mesh)
    {
      RCLCPP_ERROR(get_logger(), "Failed to load mesh from: %s", stl_resource.c_str());
      return false;
    }

    shapes::ShapeMsg mesh_msg;
    shapes::constructMsgFromShape(mesh, mesh_msg);
    delete mesh;

    shape_msgs::msg::Mesh mesh_shape = boost::get<shape_msgs::msg::Mesh>(mesh_msg);

    geometry_msgs::msg::Pose mesh_pose;
    // Stage position in meters (relative to base_link)
    // NOTE: z=0.0에 두면 로봇 베이스와 겹쳐서 "전체가 collision"처럼 보일 수 있습니다.
    mesh_pose.position.x = -0.15;  // Forward(+) / Backward(-)
    mesh_pose.position.y = 0.4;    // Left(+) / Right(-)
    mesh_pose.position.z = 0.8;    // Up(+). 예: 테이블 상판 높이
    
    // Stage orientation (quaternion)
    // IMPORTANT: quaternion은 degrees를 직접 넣는게 아니라 (x,y,z,w) 단위 quaternion이어야 합니다.
    // 회전이 필요 없으면 w=1.0, 나머지 0.0이 맞습니다.
    mesh_pose.orientation.x = 0.0;
    mesh_pose.orientation.y = 0.0;
    mesh_pose.orientation.z = 0.0;
    mesh_pose.orientation.w = 1.0;

    moveit_msgs::msg::CollisionObject co;
    co.id = object_id;
    co.header.frame_id = frame_id;
    co.meshes.push_back(mesh_shape);
    co.mesh_poses.push_back(mesh_pose);
    co.operation = moveit_msgs::msg::CollisionObject::ADD;

    // 3) Apply to planning scene
    RCLCPP_INFO(get_logger(), "Adding stage mesh to planning scene...");
    std::vector<moveit_msgs::msg::CollisionObject> collision_objects;
    collision_objects.push_back(co);
    
    bool success = planning_scene_interface_.applyCollisionObjects(collision_objects);
    
    if (success)
    {
      RCLCPP_INFO(get_logger(), "✅ Stage mesh added successfully!");
    }
    else
    {
      RCLCPP_ERROR(get_logger(), "❌ Failed to add stage mesh");
      return false;
    }

    std::this_thread::sleep_for(1s);
    
    return true;
  }

  void list_objects()
  {
    auto object_names = planning_scene_interface_.getKnownObjectNames();
    RCLCPP_INFO(get_logger(), "Objects in scene:");
    for (const auto& name : object_names)
    {
      RCLCPP_INFO(get_logger(), "  - %s", name.c_str());
    }
  }

  bool update_acm_for_stage()
  {
    RCLCPP_INFO(get_logger(), "Updating ACM to disable stage-base collisions...");
    
    // Create service clients
    auto get_scene_client = create_client<moveit_msgs::srv::GetPlanningScene>("/get_planning_scene");
    auto apply_scene_client = create_client<moveit_msgs::srv::ApplyPlanningScene>("/apply_planning_scene");
    
    if (!get_scene_client->wait_for_service(std::chrono::seconds(3)))
    {
      RCLCPP_ERROR(get_logger(), "Service /get_planning_scene not available");
      return false;
    }
    if (!apply_scene_client->wait_for_service(std::chrono::seconds(3)))
    {
      RCLCPP_ERROR(get_logger(), "Service /apply_planning_scene not available");
      return false;
    }
    
    // Get current ACM
    auto get_request = std::make_shared<moveit_msgs::srv::GetPlanningScene::Request>();
    get_request->components.components = moveit_msgs::msg::PlanningSceneComponents::ALLOWED_COLLISION_MATRIX;
    
    auto get_future = get_scene_client->async_send_request(get_request);
    if (rclcpp::spin_until_future_complete(shared_from_this(), get_future, std::chrono::seconds(5)) !=
        rclcpp::FutureReturnCode::SUCCESS)
    {
      RCLCPP_ERROR(get_logger(), "Failed to get planning scene");
      return false;
    }
    
    auto scene = get_future.get()->scene;
    auto& acm = scene.allowed_collision_matrix;
    
    // Links to disable collision with stage
    std::vector<std::string> base_links = {"base_link", "base_link_inertia", "base", "world"};
    std::string stage_name = "stage";
    
    // Find or add stage to ACM
    auto stage_it = std::find(acm.entry_names.begin(), acm.entry_names.end(), stage_name);
    size_t stage_idx;
    
    if (stage_it == acm.entry_names.end())
    {
      // Add stage to ACM
      RCLCPP_INFO(get_logger(), "Adding stage to ACM...");
      acm.entry_names.push_back(stage_name);
      stage_idx = acm.entry_names.size() - 1;
      
      // Expand existing rows
      for (auto& entry : acm.entry_values)
      {
        entry.enabled.push_back(false);
      }
      
      // Add new row for stage
      moveit_msgs::msg::AllowedCollisionEntry new_entry;
      new_entry.enabled.resize(acm.entry_names.size(), false);
      acm.entry_values.push_back(new_entry);
    }
    else
    {
      stage_idx = std::distance(acm.entry_names.begin(), stage_it);
    }
    
    // Enable collision allowance between stage and base links
    for (const auto& link : base_links)
    {
      auto link_it = std::find(acm.entry_names.begin(), acm.entry_names.end(), link);
      if (link_it != acm.entry_names.end())
      {
        size_t link_idx = std::distance(acm.entry_names.begin(), link_it);
        acm.entry_values[stage_idx].enabled[link_idx] = true;
        acm.entry_values[link_idx].enabled[stage_idx] = true;
        RCLCPP_INFO(get_logger(), "Disabled collision: %s <-> %s", stage_name.c_str(), link.c_str());
      }
    }
    
    // Apply updated ACM
    auto apply_request = std::make_shared<moveit_msgs::srv::ApplyPlanningScene::Request>();
    apply_request->scene.allowed_collision_matrix = acm;
    apply_request->scene.is_diff = true;
    
    auto apply_future = apply_scene_client->async_send_request(apply_request);
    if (rclcpp::spin_until_future_complete(shared_from_this(), apply_future, std::chrono::seconds(5)) !=
        rclcpp::FutureReturnCode::SUCCESS)
    {
      RCLCPP_ERROR(get_logger(), "Failed to apply ACM");
      return false;
    }
    
    if (apply_future.get()->success)
    {
      RCLCPP_INFO(get_logger(), "✅ ACM updated successfully!");
      return true;
    }
    else
    {
      RCLCPP_ERROR(get_logger(), "❌ Failed to update ACM");
      return false;
    }
  }

private:
  moveit::planning_interface::PlanningSceneInterface planning_scene_interface_;
};

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);
  
  auto node = std::make_shared<StageSceneSetup>();
  
  // Give MoveIt some time to initialize
  RCLCPP_INFO(node->get_logger(), "Waiting for MoveIt to initialize...");
  std::this_thread::sleep_for(2s);
  
  // Add stage mesh
  bool success = node->add_stage_mesh();
  
  if (success)
  {
    // Update ACM to disable stage-base collisions
    bool acm_success = node->update_acm_for_stage();
    
    if (!acm_success)
    {
      RCLCPP_WARN(node->get_logger(), "Stage added but ACM update failed - collision may occur");
    }
    
    // List objects
    node->list_objects();
    
    RCLCPP_INFO(node->get_logger(), "\n"
                "==================================================\n"
                "✅ Stage mesh added to planning scene!\n"
                "==================================================\n"
                "\n"
                "You can now:\n"
                "1. View the stage in RViz (MotionPlanning → Scene Objects)\n"
                "2. Run pick & place tasks\n"
                "3. Adjust object positions in code\n");
  }
  else
  {
    RCLCPP_ERROR(node->get_logger(), "Failed to add stage mesh!");
  }
  
  rclcpp::shutdown();
  return success ? 0 : 1;
}

File: src/epics_robot/scripts/add_stage_to_scene.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Add stage.stl to RViz Planning Scene
"""

import rclpy
from rclpy.node import Node
from moveit_commander import PlanningSceneInterface
from geometry_msgs.msg import PoseStamped
import time


class StageSceneSetup(Node):
    def __init__(self):
        super().__init__('stage_scene_setup')
        
        # Planning scene interface
        self.psi = PlanningSceneInterface(synchronous=True)
        
        self.get_logger().info('Stage Scene Setup node initialized')
    
    def add_stage_mesh(self):
        """Add stage.stl as mesh to planning scene"""
        
        # Clear previous objects
        self.get_logger().info('Clearing previous scene objects...')
        self.psi.remove_world_object()
        time.sleep(0.5)
        
        # Stage mesh pose (relative to base_link frame)
        stage_pose = PoseStamped()
        stage_pose.header.frame_id = "base_link"
        # Position in meters
        stage_pose.pose.position.x = 0.0  # Forward(+) / Backward(-)
        stage_pose.pose.position.y = 0.0  # Left(+) / Right(-)
        stage_pose.pose.position.z = 0.0  # Up(+) / Down(-)
        # Orientation (quaternion - w=1.0 means no rotation)
        stage_pose.pose.orientation.w = 1.0
        stage_pose.pose.orientation.x = 0.0
        stage_pose.pose.orientation.y = 0.0
        stage_pose.pose.orientation.z = 0.0
        
        # Path to STL file
        stl_path = "/home/stevek/ws/stage.stl"
        
        self.get_logger().info(f'Adding stage mesh from: {stl_path}')
        
        # Add mesh to planning scene
        # Scale factor: 0.01 converts cm to meter
        self.psi.add_mesh(
            name="stage",
            pose=stage_pose,
            filename=stl_path,
            size=(0.01, 0.01, 0.01)  # Convert cm to meter
        )
        
        time.sleep(1.0)
        self.get_logger().info('✅ Stage mesh added to planning scene!')
        
        # Optionally add a box on top of the stage for pick & place
        # box_pose = PoseStamped()
        # box_pose.header.frame_id = "base_link"
        # box_pose.pose.position.x = 0.3
        # box_pose.pose.position.y = 0.0
        # box_pose.pose.position.z = 0.1  # Above the stage
        # box_pose.pose.orientation.w = 1.0
        # 
        # self.get_logger().info('Adding box object...')
        # self.psi.add_box(
        #     name="box",
        #     pose=box_pose,
        #     size=(0.05, 0.05, 0.05)
        # )
        # 
        # time.sleep(0.5)
        # self.get_logger().info('✅ Box object added!')
    
    def list_objects(self):
        """List all objects in planning scene"""
        objects = self.psi.get_known_object_names()
        self.get_logger().info(f'Objects in scene: {objects}')
        return objects


def main(args=None):
    rclpy.init(args=args)
    
    node = StageSceneSetup()
    
    # Give MoveIt some time to initialize
    time.sleep(2.0)
    
    # Add stage mesh
    node.add_stage_mesh()
    
    # List objects
    node.list_objects()
    
    print("\n" + "="*50)
    print("✅ Stage and objects added to planning scene!")
    print("="*50)
    print("\nYou can now:")
    print("1. View the stage in RViz (MotionPlanning → Scene Objects)")
    print("2. Run pick & place tasks")
    print("3. Manually adjust object positions in code\n")
    
    rclpy.shutdown()


if __name__ == '__main__':
    main()

